core/config/stop_config.cpp

```diff
@@ -0,0 +1,154 @@
+        }
+        return params;
+    }
+};
+
+
+inline deferred_factory_parameter<stop::CriterionFactory> configure_residual(
+    const pnode& config, const registry& context, const type_descriptor& td)
+{
+    auto updated = update_type(config, td);
+    return dispatch<stop::CriterionFactory, ResidualNormConfigurer>(
+        config, context, updated,
+        make_type_selector(updated.get_value_typestr(), value_type_list()));
+}
+
+
+template <typename ValueType>
+class ImplicitResidualNormConfigurer {
+public:
+    static deferred_factory_parameter<
+        typename stop::ImplicitResidualNorm<ValueType>::Factory>
+    parse(const gko::config::pnode& config,
+          const gko::config::registry& context,
+          const gko::config::type_descriptor& td_for_child)
+    {
+        auto params = stop::ImplicitResidualNorm<ValueType>::build();
+        if (auto& obj = config.get("reduction_factor")) {
+            params.with_reduction_factor(
+                gko::config::get_value<remove_complex<ValueType>>(obj));
+        }
+        if (auto& obj = config.get("baseline")) {
+            params.with_baseline(get_mode(obj.get_string()));
+        }
+        return params;
+    }
+};
+
+
+inline deferred_factory_parameter<stop::CriterionFactory>
+configure_implicit_residual(const pnode& config, const registry& context,
+                            const type_descriptor& td)
+{
+    auto updated = update_type(config, td);
+    return dispatch<stop::CriterionFactory, ImplicitResidualNormConfigurer>(
+        config, context, updated,
+        make_type_selector(updated.get_value_typestr(), value_type_list()));
+}
+
+
+template <>
+deferred_factory_parameter<const stop::CriterionFactory>
+get_factory<const stop::CriterionFactory>(const pnode& config,
+                                          const registry& context,
+                                          const type_descriptor& td)
+{
+    deferred_factory_parameter<const stop::CriterionFactory> ptr;
+    if (config.get_tag() == pnode::tag_t::string) {
+        return detail::registry_accessor::get_data<stop::CriterionFactory>(
+            context, config.get_string());
+    } else if (config.get_tag() == pnode::tag_t::map) {
+        static std::map<std::string,
+                        std::function<deferred_factory_parameter<
+                            gko::stop::CriterionFactory>(
+                            const pnode&, const registry&, type_descriptor)>>
+            criterion_map{
+                {{"stop::Time", configure_time},
+                 {"stop::Iteration", configure_iter},
+                 {"stop::ResidualNorm", configure_residual},
+                 {"stop::ImplicitResidualNorm", configure_implicit_residual}}};
+        return criterion_map.at(config.get("type").get_string())(config,
+                                                                 context, td);
+    }
+    GKO_THROW_IF_INVALID(!ptr.is_empty(), "Parse get nullptr in the end");
+    return ptr;
+}
```

Comment:
Why is this in here and not in `config_helper.cpp`? The definition is in `config_helper.hpp`.

Reply:
I put this here because I do not need to declare the function in another header.

Reply:
But you do declare it in `config_helper.hpp`, which is why I would suggest to put it into `config_helper.cpp`:
https://github.com/ginkgo-project/ginkgo/blob/d4bc0a987b0b0571a3b0a6dbc494cec67b436572/core/config/config_helper.hpp#L85